File: ProjEuler6.py
#!/usr/bin/python
# https://projecteuler.net/problem=6

def method1():
    '''Algebraically, the difference between the square of the sum and sum of squares is
       twice the sum of the pairwise produces between distinct summands.'''  
    total1 = 0

    for i in range(1,101):
        for j in range(i+1,101):
	    total1 += i*j
    total1 *= 2
    print "The answer via method 1 is: %d"%(total1,)

def method2():
    '''This is just the naive computation, not being clever.'''
    sqSum = sum(range(1,101))**2
    sumSq = sum(i**2 for i in range(1,101))
    total2 = sqSum - sumSq
    print "The answer via method 2 is: %d"%(total2,)

if __name__ == '__main__':
    import timeit
    print timeit.timeit("method1()", setup="from __main__ import method1",number=1)
    print timeit.timeit("method2()", setup="from __main__ import method2",number=1)

# it turns out method2 is faster by two orders of magnitude.
# Analysis of method1:
# two operations repeated 

File: ProjEuler6.pl
#!/usr/bin/perl
# https://projecteuler.net/problem=6

# Find sum squared minus sum of squares for first 100 natural numbers

use warnings;
use strict;
use List::Util;

my ($sum1,$sum2,$i)=(0,0,0);

$sum1 = &List::Util::sum(1..100);
$sum1 = $sum1**2;

my @array = map {$_**2} (1..100);
$sum2 = &List::Util::sum(@array);

my $answer = $sum1 - $sum2;

print "The answer is: $answer\n";

exit;
